# 00. Table of content
### 01. Importing Librabries & Dataframes
### 02. Data Wrangling
### 03. Check Dataframes for Upcoming Merge
### 04. Exporting Cleaned Data

# 01. Importing Libraries & Dataframes

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Create path
path = r'C:\Users\krist\22.01.2024 Instacart Basket Analysis'

In [3]:
# Import customer dataframe
df_customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Import ords_prods_merge dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))


# 02. Data Wrangling

In [5]:
# Check shape
df_customer.shape

(206209, 10)

In [6]:
# Check data types
df_customer.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [7]:
# Check Head
df_customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [8]:
# Rename column First Name
df_customer = df_customer.rename(columns = {'First Name':'first_name'})

In [9]:
# Check
df_customer.dtypes

user_id          int64
first_name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [10]:
# Rename column Surname
df_customer = df_customer.rename(columns = {'Surnam':'last_name'})

In [11]:
# Rename column Gender
df_customer = df_customer.rename(columns = {'Gender':'gender'})

In [12]:
# Rename column STATE
df_customer = df_customer.rename(columns = {'STATE':'state'})

In [13]:
# Rename column Age
df_customer = df_customer.rename(columns = {'Age':'age'})

In [14]:
# Check
df_customer.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [15]:
# Change datatype for user_id
df_customer['user_id'] = df_customer['user_id'].astype('str')

In [16]:
# Describe dataframe
df_customer.describe()

age   n_dependants         income
count  206209.000000  206209.000000  206209.000000
mean       49.501646       1.499823   94632.852548
std        18.480962       1.118433   42473.786988
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59874.000000
50%        49.000000       1.000000   93547.000000
75%        66.000000       3.000000  124244.000000
max        81.000000       3.000000  593901.000000

Comments: Age range, Dependants range and income range look okay (max income is potentially high, but not impossible)

In [17]:
# Find missing values
df_customer.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

Comment: There are 11259 rows where the first_name is missing. I see this variable as not crucial to the upcoming analysis and also wouldn't impute it with another value.

In [18]:
# Check on missing values
df_nan = df_customer[df_customer['first_name'].isnull() == True]

In [19]:
df_nan.head()

user_id first_name last_name  gender      state  age date_joined  \
53    76659        NaN   Gilbert    Male   Colorado   26    1/1/2017   
73    13738        NaN     Frost  Female  Louisiana   39    1/1/2017   
82    89996        NaN    Dawson  Female     Oregon   52    1/1/2017   
99    96166        NaN   Oconnor    Male   Oklahoma   51    1/1/2017   
105   29778        NaN    Dawson  Female       Utah   63    1/1/2017   

     n_dependants fam_status  income  
53              2    married   41709  
73              0     single   82518  
82              3    married  117099  
99              1    married  155673  
105             3    married  151819

In [20]:
# Find duplicates
df_dups = df_customer[df_customer.duplicated()]

In [21]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

Comment: No duplicates found - no action required

In [22]:
# Check tail
df_customer.tail()

user_id first_name last_name  gender           state  age date_joined  \
206204  168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205   49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206  135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207   81095       Rose   Rollins  Female      California   27    4/1/2020   
206208   80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095

# 03. Check Dataframes for Upcoming Merge

In [23]:
# Check ords_prods_merge dataframe for combineable columns
ords_prods_merge.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                     NaN        196                  1          0   
1                    15.0        196                  1          1   
2                    21.0        196                  1          1   
3                    29.0        196                  1          1   
4                    28.0        196                  1          1   

  product_name  ... prices     busiest_day    busiest_days  \
0         Soda  ...    9.0  regularly busy  regularly busy   
1         Soda  ...    9.0  regularly busy      least busy   
2         Soda  ...    9.0  regularly busy      least busy   
3         Soda  ...    9.0      least busy      least busy   
4         Soda  ...    9.0      least busy      least busy   

  busiest_period_of_day max_order  loyalty_flag  mean_product_price  \
0        average orders        10  new customer            6.367797   
1        average orders        10  new customer            6.367797   
2           most orders        10  new customer            6.367797   
3        average orders        10  new customer            6.367797   
4           most orders        10  new customer            6.367797   

  spending_flag  median_days_since_prior_order   order_frequency_flag  
0   low spender                           20.5  non-frequent customer  
1   low spender                           20.5  non-frequent customer  
2   low spender                           20.5  non-frequent customer  
3   low spender                           20.5  non-frequent customer  
4   low spender                           20.5  non-frequent customer  

[5 rows x 22 columns]

In [24]:
# Check for matching datatypes
df_customer.dtypes

user_id         object
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [25]:
# Check for matching datatypes
ords_prods_merge.dtypes

order_id                          object
user_id                           object
order_number                       int64
order_day_of_week                  int64
order_hour_of_day                  int64
days_since_prior_order           float64
product_id                        object
add_to_cart_order                  int64
reordered                          int64
product_name                      object
aisle_id                          object
department_id                     object
prices                           float64
busiest_day                       object
busiest_days                      object
busiest_period_of_day             object
max_order                          int64
loyalty_flag                      object
mean_product_price               float64
spending_flag                     object
median_days_since_prior_order    float64
order_frequency_flag              object
dtype: object

In [26]:
# Check for shape
df_customer.shape

(206209, 10)

In [27]:
# Check for shape
ords_prods_merge.shape

(32404859, 22)

# 04. Exporting Cleaned Data

In [29]:
# Export data to pkl

df_customer.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_customer.pkl'))